clang-tools-extra/clang-tidy/readability/RedundantInlineSpecifierCheck.cpp

```diff
@@ -0,0 +1,113 @@
+//===--- RedundantInlineSpecifierCheck.cpp - clang-tidy--------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "RedundantInlineSpecifierCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Decl.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/DeclTemplate.h"
+#include "clang/AST/ExprCXX.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Basic/SourceManager.h"
+#include "clang/Lex/Token.h"
+
+#include "../utils/LexerUtils.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+AST_POLYMORPHIC_MATCHER(isInlineSpecified,
+                        AST_POLYMORPHIC_SUPPORTED_TYPES(FunctionDecl,
+                                                        VarDecl)) {
+  if (const auto *FD = dyn_cast<FunctionDecl>(&Node))
+    return FD->isInlineSpecified();
+  if (const auto *VD = dyn_cast<VarDecl>(&Node))
+    return VD->isInlineSpecified();
+  llvm_unreachable("Not a valid polymorphic type");
+}
+
+static std::optional<SourceLocation>
+getInlineTokenLocation(SourceRange RangeLocation, const SourceManager &Sources,
+                       const LangOptions &LangOpts) {
+  SourceLocation Loc = RangeLocation.getBegin();
+  if (Loc.isMacroID())
+    return std::nullopt;
+
+  Token FirstToken;
+  Lexer::getRawToken(Loc, FirstToken, Sources, LangOpts, true);
+  std::optional<Token> CurrentToken = FirstToken;
+  while (CurrentToken && CurrentToken->getLocation() < RangeLocation.getEnd() &&
+         CurrentToken->isNot(tok::eof)) {
+    if (CurrentToken->is(tok::raw_identifier) &&
+        CurrentToken->getRawIdentifier() == "inline")
+      return CurrentToken->getLocation();
+
+    CurrentToken =
+        Lexer::findNextToken(CurrentToken->getLocation(), Sources, LangOpts);
+  }
+  return std::nullopt;
+}
+
+void RedundantInlineSpecifierCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      functionDecl(isInlineSpecified(),
+                   anyOf(isConstexpr(), isDeleted(), isDefaulted(),
+                         isInAnonymousNamespace(),
```

Comment:
To be honest I'm not so sure about anonymous namespace, static keyword, and templates.
Looks like in those cases those functions are inline only due to optimizations, not because they "inline".

This is more visible in GCC when you use attribute:
```inline int __attribute__((always_inline))  fn2(int i) ```
if you change inline to static, or put this in anonymous namespace you get warning:
`warning: 'always_inline' function might not be inlinable [-Wattributes]`

Best would be add "StrictMode" option to check, and when set to true then emit warning also for templates and static and anonymous namespace, and when set to false, emit it only for class members and constexpr.
Please investigate it more.

Reply:
According to [cppref](https://en.cppreference.com/w/cpp/language/inline), static and anon namespace variables are implicitly inline when they are also constexpr.

It doesn't say anything about static functions, probably because a static, inline function makes no sense.

Statically declared functions are only visible in their TU, so inline would be redundant. Inline only allows redefinition across multiple TUs. Not sure if modules change any of that.

So technically it's pointless, not implicit. Same with templates.

Reply:
I asked around and it seems like inline for templates and static functions also adds a compiler hint for inline optimization as a secondary effect of the inline keyword.

https://clang.godbolt.org/z/8q3sb5Ezn

Reply:
Adding a parameter to differentiate "implicit" from "redundant" inline declarations seems like a good idea. Some users might be using it to hint the optimizer and might want to keep it if it's not implicit.

I will add the parameter "StrictMode" suggested by @PiotrZSL which will be an extra condition to flag the following cases if they are not constexpr:
- templates
- functions declared in an anonymous namespace 
- functions marked as static